trieste/models/gpflow/models.py

```diff
@@ -1647,3 +1647,367 @@ def covariance_with_top_fidelity(self, query_points: TensorType) -> TensorType:
+        :return: mean: The mean at query_points with shape [..., N, P],
+            and var: The variance at query_points with shape [..., N, P]
+        """
+        check_and_extract_fidelity_query_points(query_points, max_fidelity=self.num_fidelities - 1)
+
+        sample_mean, sample_var = self._sample_mean_and_var_at_fidelities(
+            query_points
+        )  # [..., N, 1, S], [..., N, 1, S]
+        variance = tf.reduce_mean(sample_var, axis=-1) + tf.math.reduce_variance(
+            sample_mean, axis=-1
+        )
+        mean = tf.reduce_mean(sample_mean, axis=-1)
+        return mean, variance
+
+    def predict_y(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        """
+        Predict the marginal mean and variance at `query_points` including observation noise
+
+        :param query_points: Query points with shape [..., N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :return: mean: The mean at query_points with shape [N, P],
+            and var: The variance at query_points with shape [N, P]
+        """
+        _, query_points_fidelity_col = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )
+
+        f_mean, f_var = self.predict(query_points)
+
+        # Get fidelity 0 observation noise
+        observation_noise = (
+            tf.ones_like(query_points_fidelity_col)
+            * self.fidelity_models[0].get_observation_noise()
+        )
+
+        for fidelity in range(1, self.num_fidelities):
+
+            fidelity_observation_noise = self.fidelity_models[fidelity].get_observation_noise()
+
+            mask = query_points_fidelity_col >= fidelity
+
+            observation_noise = tf.where(mask, fidelity_observation_noise, observation_noise)
+        return f_mean, f_var + observation_noise
+
+    def _sample_mean_and_var_at_fidelities(
+        self, query_points: TensorType
+    ) -> tuple[TensorType, TensorType]:
+        """
+        Draw `num_monte_carlo_samples` samples of mean and variance from the model at the fidelities
+        passed in the final column of the query points.
+
+        :param query_points:  Query points with shape [..., N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :return: sample_mean: Samples of the mean at the query points with shape [..., N, 1, S]
+            and sample_var: Samples of the variance at the query points with shape [..., N, 1, S]
+        """
+
+        (
+            query_points_wo_fidelity,
+            query_points_fidelity_col,
+        ) = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )  # [..., N, D], [..., N, 1]
+
+        sample_mean, sample_var = self.fidelity_models[0].predict(
```

Comment:
naive question: is this what we want to do? i.e. use a single sample mean and var and duplicate it? or would we ideally want S values of `sample_mean` and `sample_var`?

Reply:
Ah yes, sorry, I think the naming is bad here